src/config.c

```diff
@@ -0,0 +1,149 @@
+    return ((RM_ConfigFloat *)privdata)->str_value;
+}
+
+static int setIntegerValue(const char *name, RedisModuleString *value, void *privdata,
+                           RedisModuleString **err) {
+    RM_ConfigInteger *config = privdata;
+    long long new_val;
+    if (RedisModule_StringToLongLong(value, &new_val) != REDISMODULE_OK) {
+        *err = RedisModule_CreateStringPrintf(NULL, "Invalid value for `%s`", name);
+        return REDISMODULE_ERR;
+    }
+    if (!isIntegerConfigValid(new_val, *config)) {
+        *err =
+            RedisModule_CreateStringPrintf(NULL, "Value for `%s` must be in the range [%lld, %lld]",
+                                           name, config->min, config->max);
+        return REDISMODULE_ERR;
+    }
+    config->value = new_val;
+    RedisModule_FreeString(NULL, config->str_value);
+    config->str_value = RedisModule_HoldString(NULL, value);
+
+    if (!RM_ConfigStrCaseCmp(name, cf_bucket_size)) {
+        rm_config.cf_initial_size.min = new_val * 2;
+    } else if (!RM_ConfigStrCaseCmp(name, cf_initial_size)) {
+        rm_config.cf_bucket_size.max = new_val / 2 < 255 ? new_val / 2 : 255;
+    }
+
+    return REDISMODULE_OK;
+}
+static RedisModuleString *getIntegerValue(const char *name, void *privdata) {
+    return ((RM_ConfigInteger *)privdata)->str_value;
+}
+
+#define registerConfigVar(config)                                                                  \
+    do {                                                                                           \
+        const char *name = RM_ConfigOptionToString(config);                                        \
+        rm_config.config.str_value = RedisModule_CreateStringPrintf(                               \
+            NULL, _Generic(rm_config.config.value, long long: "%lld", double: "%f"),               \
+            rm_config.config.value);                                                               \
+        const char *default_val = RedisModule_StringPtrLen(rm_config.config.str_value, NULL);      \
+        if (RedisModule_RegisterStringConfig(ctx, name, default_val,                               \
+                                             REDISMODULE_CONFIG_UNPREFIXED,                        \
+                                             _Generic(rm_config.config.value,                      \
+                                             long long: getIntegerValue,                           \
+                                             double: getFloatValue),                               \
+                                             _Generic(rm_config.config.value,                      \
+                                             long long: setIntegerValue,                           \
+                                             double: setFloatValue),                               \
+                                             NULL, &rm_config.config) != REDISMODULE_OK) {         \
```

Comment:
Can we have a static function for this code?

```c
        const char *name = RM_ConfigOptionToString(config);                                        \
        rm_config.config.str_value = RedisModule_CreateStringPrintf(                               \
            NULL, _Generic(rm_config.config.value, long long: "%lld", double: "%f"),               \
            rm_config.config.value);                                                               \
        const char *default_val = RedisModule_StringPtrLen(rm_config.config.str_value, NULL);      \
        if (RedisModule_RegisterStringConfig(ctx, name, default_val,                               \
                                             REDISMODULE_CONFIG_UNPREFIXED,                        \
                                             _Generic(rm_config.config.value,                      \
                                             long long: getIntegerValue,                           \
                                             double: getFloatValue),                               \
                                             _Generic(rm_config.config.value,                      \
                                             long long: setIntegerValue,                           \
                                             double: setFloatValue),                               \
                                             NULL, &rm_config.config) != REDISMODULE_OK) { 
```

And within this macro, just invoke it within the same `do while(0)` loop. In my opinion, this would make the code a little more readable and concise. This function, though, would be better to have the `rm_config` passed as an argument as well, to make it purer, but this shouldn't be a problem to do from within this macro.

Reply:
You can still extract this info a function and have the if and the return statements just as you described.

Reply:
unfortunately not. because there is a lot of token pasting in the macro, which cannot be done at runtime. `rm_config.config.value` expands to `rm_config.<actual-config-variable-name>.value`, etc